### PR TITLE
WIP development, mostly name indexing structure

### PR DESCRIPTION
**Pull request by markATavail**

### CODE REVIEW

File: avail/src/main/kotlin/avail/persistence/cache/record/NameInModule.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+data class NameInModule constructor(
+	val moduleName: String,
+	val atomName: String
+): Comparable<NameInModule>
+{
+	override fun compareTo(other: NameInModule): Int =
+		moduleName.compareTo(other.moduleName)
+			.ifZero { atomName.compareTo(other.atomName) }
+
+	fun write(
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NameInModule.kt:55]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+import avail.persistence.cache.record.PhrasePathRecord.PhraseNode
+import avail.utility.decodeString
+import avail.utility.sizedString
+import avail.utility.structures.BloomFilter
+import avail.utility.unvlqInt
+import avail.utility.vlq
+import java.io.DataInputStream
+import java.io.DataOutputStream
+import java.io.IOException
+/**
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:52]: Missing blank line.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+class NamesIndex
+{
+	/**
+	 * A [Map] from each occurring [NameInModule] to the [NameOccurrences] that
+	 * describe where and how the name is used in this module.
+	 */
+	val occurrences: MutableMap<NameInModule, NameOccurrences>
+
+	/**
+	 * Iff this index is for a package, this is a bloom filter that can be used
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:66]: Capitalize Bloom.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+	 *   A list of the name's [Definition]s present in this module.
+	 * @property usages
+	 *   A list of places that the name has [Usage]s within this module.
+	 */
+	data class NameOccurrences(
+		val declarations: MutableList<Declaration>,
+		val definitions: MutableList<Definition>,
+		val usages: MutableList<Usage>)
+	{
+		fun write(
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:91]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+			}
+			binaryStream.vlq(usages.size)
+			usages.forEach { usage ->
+				usage.write(binaryStream)
+			}
+		}
+
+		/**
+		 * Reconstruct a [NameOccurrences] from a stream, using the already
+		 * constructed lists of module names and atom names
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:112]: Incomplete comment; missing full stop, `@param` tags.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+	 *   An index into the canonically ordered [PhraseNode]s in the
+	 *   [PhrasePathRecord] for this module compilation.  That [PhraseNode] is
+	 *   enough information to identify a region of the module to highlight
+	 *   or select to identify the declaration of the name.
+	 */
+	data class Declaration(
+		val alias: NameInModule?,
+		val phraseIndex: Int)
+	{
+		fun write(
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:147]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+		{
+			binaryStream.writeBoolean(alias != null)
+			alias?.run { write(binaryStream, moduleNumbering, nameNumbering) }
+			binaryStream.vlq(phraseIndex)
+		}
+
+		/**
+		 * Reconstruct a [NameInModule] from a stream, using the already constructed
+		 * lists of module names and atom names
+		 */
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:160]: Incomplete comment; missing full stop, @param tags.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+	 *   The [DefinitionType] identifying the nature of this definition.
+	 * @property manifestIndex
+	 *   An index into the [ManifestRecord] for the current module, indicating
+	 *   which manifest entry is responsible for this definition.
+	 */
+	data class Definition(
+		val definitionType: DefinitionType,
+		val manifestIndex: Int)
+	{
+		fun write(binaryStream: DataOutputStream)
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:187]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+		{
+			binaryStream.vlq(definitionType.ordinal)
+			binaryStream.vlq(manifestIndex)
+		}
+
+		constructor(
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:193]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+	 *   An index into the canonically ordered [PhraseNode]s in the
+	 *   [PhrasePathRecord] for this module compilation.  That [PhraseNode] is
+	 *   enough information to identify a region of the module to highlight
+	 *   or select to identify the usage of the name.
+	 */
+	data class Usage(
+		val usageType: UsageType,
+		val phraseIndex: Int)
+	{
+		fun write(binaryStream: DataOutputStream)
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:254]: Missing comment.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+		{
+			binaryStream.vlq(usageType.ordinal)
+			binaryStream.vlq(phraseIndex)
+		}
+
+		/**
+		 * Reconstruct a [Usage] from a stream.
+		 */
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:262]: Incomplete comment. Missing `@param`.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+	}
+
+	/**
+	 * Add information about a [Usage].
+	 *
+	 * @param nameInModule
+	 *   The [NameInModule] being used.
+	 * @param
+	 *   The nature of the usage of the [nameInModule].
+	 * @param
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:378]: Kill.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+		withOccurrences: (NameOccurrences) -> Unit)
+	{
+		occurrences[nameToFind]?.let(withOccurrences)
+	}
+
+	/**
+	 * Output this [NamesIndex] onto a [DataOutputStream], so that Anvil can
+	 * later access it to determine what names are used, and in what ways,
+	 * inside this module.
+	 */
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:433]: Missing `@param`.

File: avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+
+	/**
+	 * Construct a [NamesIndex] from the provided data.
+	 *
+	 * @param occurrences
+	 *   A [MutableMap] from each occurring [NameInModule] to the
+	 *   [NameOccurrences] that describe where and how the name is used in this
+	 *   module.
+	 * @param bloomFilterIfPackage
+	 *   Iff this index is for a package, this is a bloom filter that can be
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/persistence/cache/record/NamesIndex.kt:526]: Capitalize "bloom".

File: avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt [hunk 15]
```diff
@@ -185,13 +214,58 @@ class BloomFilter
...
 		assert (hashCount == filter.hashCount)
 		for (i in array.indices)
 		{
 			array[i] = array[i] or filter.array[i]
 		}
 	}
+
+	/**
+	 * Write this [BloomFilter] onto the given stream.
+	 */
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt:229]: Missing `@param`.

File: avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt [hunk 16]
```diff
@@ -185,13 +214,58 @@ class BloomFilter
...
+	fun write(binaryStream: DataOutputStream)
+	{
+		binaryStream.vlq(bitCount)
+		binaryStream.vlq(hashCount)
+		array.forEach(binaryStream::writeLong)
+	}
+
+	constructor(
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt:237]: Missing comment.

File: avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt [hunk 17]
```diff
@@ -185,13 +214,58 @@ class BloomFilter
...
+		binaryStream.unvlqInt())
+	{
+		array.indices.forEach {
+			array[it] = binaryStream.unvlqLong()
+		}
+	}
+
+	companion object
+	{
+
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/utility/structures/BloomFilter.kt:250]: Remove blank line.

**Review by toddATavail — CHANGES_REQUESTED** [review]
The logic looks good, and the abstractions are nice. I didn't find any bugs by eye, but I did find some rough places in the documentation, mostly due to completely missing elements and tags. Please address these and then I'll merge.

**Comment by markATavail** [comment]
Whoops, there was a bug in `BloomFilter(DataInputStream)`, where it was unvlq-ing the longs instead of reading them directly, making it inconsistent with `write`.

**Comment by markATavail** [comment]
(accidentally merged before pushing updates.  Re-merged manually.)

**Pull request merged**